### ISSUE TITLE
Make a CDN for Serelex images

### ISSUE THREAD

**Issue by alexanderpanchenko** [suhasbhairav/JuliaMachineLearning#5]
Your goal will be to create an efficient content delivery network (CDN) for a big database of images. This web service would provide an access to a collection of images via a RESTful http endpoint by word. For instance, for work “cat” you should provide one image of the cat.

Download database of images: http://panchenko.me/data/serelex/lineart.tgz

Create a RESTful HTTP server that would provide one image by one word:
• Word: “python”, Endpoint: http://my-image-server.com/python returns a url to image such as
http://cdn.my-image-server.com/python.jpg .

Use the Windows Azure Storage: 

pip install azure
sudo apt-get install libffi-dev python3-dev libssl-dev

Use the following code to upload test images:

```
#!/usr/bin/python

import os
import argparse
from azure.storage import BlobService

container = 'images'
account_name='serelex'
account_key=''


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='Upload all files from the specific directory to the Azure Storage Container')
    parser.add_argument("directory", help='Directory with the files')

    args = parser.parse_args()  

    blob_service = BlobService(account_name=account_name, account_key=account_key)

    files = os.listdir(args.directory)

    for i,f in enumerate(files):
        fullname = os.path.join(args.directory, f)
        blob_service.put_block_blob_from_path(container, f, fullname)

        if i % 50 == 0:
            print('Uploaded', i, 'files:', f)

```

**Comment by suhasbhairav** [comment]
I have initiated the uploading of images. I'm not sure how long this might take since 500MB needs to be uploaded.

**Comment by suhasbhairav** [comment]
After uploading 1250 images, it threw an error: ssl.sslError: The write operation timed out. I'm not sure why this occurred. I am still trying to find out the root cause for the error.

**Comment by alexanderpanchenko** [comment]
try to upload from the server frink, the internet connection should be faster and more stable

**Comment by suhasbhairav** [comment]
Can I request Mr.Eugen Ruppert to install azure and execute the next command after that on the frink machine? I am getting permission denied error while installing the same.

**Comment by alexanderpanchenko** [comment]
no need to do it

you can install any python packages locally with 

```
pip install --user
```

**Comment by suhasbhairav** [comment]
I tried by running pip install --user azure and it threw the following error:
Command /usr/bin/python2.6 -c "import setuptools;**file**='/srv/home/suhas/machinelearningfiles/build/cryptography/setup.py';exec(compile(open(**file**).read().replace('\r\n', '\n'), **file**, 'exec'))" install --single-version-externally-managed --record /tmp/pip-oWoSXj-record/install-record.txt --user failed with error code 1 in /srv/home/suhas/machinelearningfiles/build/cryptography

So I tried to install it by running setup.py inside the cryptography folder and then it threw the following error:
No module named pkg_resources.

I tried running setup.py inside setuptools of build folder but still in vain as it threw me a permission denied error.

**Comment by alexanderpanchenko** [comment]
you need to install pkg_resources first 

```
pip install --user pkg_resources
```

**Comment by suhasbhairav** [comment]
I did try that. It said it cannot find pkg_resources under pip. After googling for the error, I came to know that one has to upgrade setuptools (http://stackoverflow.com/questions/7446187/no-module-named-pkg-resources and http://stackoverflow.com/questions/22148144/python-importerror-no-module-named-pkg-resources). Ran the installation of setup tools using pip install --upgrade --user setuptools. Then I am getting permission denied error because it wants to rewrite setuptools folder under dist-packages of python installation and I do not have permissions for it. I'm trying if there are other ways to do so. I have been trying to fix this from a few days. Hopefully, there should be a way around this.

**Comment by alexanderpanchenko** [comment]
can be the problem indeed

please write to Eugene with a request to update setuptools (specify exact commands)
it is actually quite good thing to update pip in any case

**Comment by suhasbhairav** [comment]
All the images (3656 to be precise) have been uploaded to the server. It worked on the university network without any interruption. 

When I tried accessing my-image-server.com, it prompted for the username and password which I entered as mentioned in the script. But still it is not allowing me to view it. It says "Sorry, Please try again". Does that mean it can be accessed only by using a webservice (that I'm supposed to create) or should it work as it is at the moment?

Note:- I guess there won't be any need for me to send a mail to Eugene now.

**Comment by alexanderpanchenko** [comment]
my-image-server.com is only an example

**Comment by alexanderpanchenko** [comment]
available here
https://serelex.blob.core.windows.net/images/cabal.jpg

**Comment by suhasbhairav** [comment]
Okay. So I would need to write a REST HTTP server that connects to https://serelex.blob.core.windows.net to display an image for a query like word = python.
Is my understanding correct?

**Comment by alexanderpanchenko** [comment]
wait a bit for now, just switch to other tasks

**Comment by alexanderpanchenko** [comment]
The images are now available on frink:/home/panchenko/images/img-clipart/ 

1) Clean image names. 

Write a script for name cleansing e.g. 

```
acupunture.jpg\\\?oh\\\=996dc26fcbaec37faa1d88a4ffbaf0ca\\\&oe\\\=55F9DE4E\\\&__gda__\\\=1442098114_9674778d0dbacbe92265f06116eee2a1
```

have to be

```
acupunture.jpg
```

2) Upload all images to the Windows Azure server. 

Let me know when upload is done.

**Comment by suhasbhairav** [comment]
There are a few non-image files. For example there are few javascript files. Should they also be uploaded?

**Comment by alexanderpanchenko** [comment]
actually it is needed to filter the files slightly (see above). just delete the js files and other irrelevant files

**Comment by suhasbhairav** [comment]
I'm done with the python script to clean image file names but I'm not able to upload it on the server due to permission restrictions like last time. 

For the previous upload, the folder size was very less. So I was able to upload it from my local machine . But now I cannot do that from my local machine as there are too many files. But even the pip command doesn't work as it has to write to the main python directory while adding packages just like last time.

pip install azure
sudo apt-get install libffi-dev python3-dev libssl-dev

The above commands need to be executed by someone with admin privileges for me to proceed forward.

**Comment by alexanderpanchenko** [comment]
please write to Eugen Ruppert: describe the problem and ask him to make an update

**Comment by suhasbhairav** [comment]
azure module is already installed in python 2.6 on the frink machine. I was trying on python 2.7 to install it . I tried running the script using 2.6 version. I encountered the following error:
File "/home/suhas/.local/lib/python2.6/site-packages/azure/storage/**init**.py", line 1193
    name.lower() : value for name, value in request.headers if value

I came to know this is a version specific issue of azure with respect to python 2.6. That has been fixed in the following link.
https://github.com/Azure/azure-sdk-for-python/issues/395

But I believe the azure package installed on the frink machine for version 2.6 does not contain this fix.
